sdk/core/System.ClientModel/src/Message/BinaryContent.cs

```diff
@@ -0,0 +1,179 @@
+    /// Writes contents of this object to an instance of <see cref="Stream"/>.
+    /// </summary>
+    /// <param name="stream">The stream to write to.</param>
+    /// <param name="cancellationToken">To cancellation token to use.</param>
+    public abstract void WriteTo(Stream stream, CancellationToken cancellationToken);
+
+    /// <inheritdoc/>
+    public abstract void Dispose();
+
+    private sealed class BinaryDataBinaryContent : BinaryContent
+    {
+        private readonly ReadOnlyMemory<byte> _bytes;
+
+        public BinaryDataBinaryContent(ReadOnlyMemory<byte> bytes)
+        {
+            _bytes = bytes;
+        }
+
+        public override bool TryComputeLength(out long length)
+        {
+            length = _bytes.Length;
+            return true;
+        }
+
+        public override void WriteTo(Stream stream, CancellationToken cancellation)
+        {
+            byte[] buffer = _bytes.ToArray();
+            stream.Write(buffer, 0, buffer.Length);
+        }
+
+        public override async Task WriteToAsync(Stream stream, CancellationToken cancellation)
+            => await stream.WriteAsync(_bytes, cancellation).ConfigureAwait(false);
+
+        public override void Dispose() { }
+    }
+
+    private sealed class ModelBinaryContent<T> : BinaryContent where T : IPersistableModel<T>
+    {
+        private readonly T _model;
+        private readonly ModelReaderWriterOptions _options;
+
+        // Used when _model is an IJsonModel
+        private ModelWriter? _writer;
+
+        // Used when _model is an IModel
+        private BinaryData? _data;
+
+        public ModelBinaryContent(T model, ModelReaderWriterOptions options)
+        {
+            _model = model;
+            _options = options;
+        }
+
+        private ModelWriter Writer
+        {
+            get
+            {
+                if (_model is not IJsonModel<T> jsonModel)
+                {
+                    throw new InvalidOperationException("Cannot use Writer with non-IJsonModel model type.");
+                }
+
+                _writer ??= new ModelWriter((IJsonModel<object>)jsonModel, _options);
+                return _writer;
+            }
+        }
+
+        private BinaryData Data
+        {
+            get
+            {
+                if (_model is IJsonModel<T> && _options.Format == "J")
+                {
+                    throw new InvalidOperationException("Should use ModelWriter instead of _model.Write with IJsonModel.");
+                }
+
+                _data ??= _model.Write(_options);
+                return _data;
+            }
+        }
+
+        public override bool TryComputeLength(out long length)
+        {
+            if (_model is IJsonModel<T> && _options.Format == "J")
+            {
+                return Writer.TryComputeLength(out length);
+            }
+
+            length = Data.ToMemory().Length;
+            return true;
+        }
+
+#if NETFRAMEWORK || NETSTANDARD2_0
```

Comment:
I would like to avoid ifdefs like this in ClinrModel as much as possible. It will be quite common for us to write BinaryData to various streams and so I think we should: 1) create one helper method in this package that writes BinaryData to stream, 2) work with the BCL team to potentially move the helper to BinaryData. 

In our implementation, we should call ToMemory() and then TryGetArray on the retrieved memory, and then write the array to the stream. If TryGetArray returns false, we would just call ToArray. 

I think TryGetArray will never return false in this (ClientModel) code paths as we always use arrays as the backing store for serialized models.

Reply:
Tracking this and other ifdefs with https://github.com/Azure/azure-sdk-for-net/issues/41221